### PR TITLE
Multiple load dimensions

### LINKED ISSUE #458: Multiple load dimensions
#### Is your feature request related to a problem? Please describe

Right now we consider a single capacity constraint, which might resemble weight. We should probably be able to (simultaneously) support constraints on the following:
- Size (width, height, depth)
- Weight (kg, litre)
- \# items (bags, packages, boxes)

Size restrictions can quickly turn into a bin packing problem, which is pretty hard (see some of the references below). I do not think we need to add support for bin packing via this issue.

#### Describe the solution you'd like

Some support for additional loading constraints.

#### Describe alternatives you have considered

N/A

#### Additional context

Came up in a conversation I had today and is something we discussed before, I think. I'm writing this out here so I do not forget about it.

### PR DESCRIPTION
**Pull request by N-Wouda**
Adds support for multidimensional load (client deliveries, client pickups, vehicle capacities). This PR:

- [x] Closes #458.
- [x] Adds and passes relevant tests.
- [x] Has well-formatted code and documentation.

From #651 by @wouterkool:

Key implementation details:
- ~`ProblemData`: the delivery and pickup amount of a `Client` and the capacity of a `VehicleType` is now a vector instead of a scalar. All delivery, pickup and capacity vectors should have the same size in a problem instance.~
- ~When creating these objects, one can either pass a list of values, or a scalar which will automatically be converted to a vector with a single element (defined conversion in `bindings.h`).~
- ~In the Python bindings, for backwards compatibility, the `Client.delivery`, `Client.pickup` and `VehicleType.capacity` properties remain scalars that return the values for the first load dimension. For additional load dimensions, the `get_(delivery|pickup|capacity)` methods can be used. Other methods that already used a getter function now have an additional argument to specify the load dimension. This is backwards compatible but not ideal, see 'open questions' below.~
- ~`LoadSegment`: is per load dimension, so the attributes of `LoadSegment` remain scalars.~ 
- ~`search/Route`: for each `Node` in the `Route` a vector of `LoadSegment`'s is stored for the caches `loadAt/loadAfter/loadBefore` => each `LoadSegment` corresponds to a single load dimension.~
- ~`CostEvaluator`: excess load penalty calculations are done per load dimension. This avoids the need for intermediate vector allocations.~ Currently the penalty used is the same across all load dimensions, assuming these are scaled appropriately.
- ~Fleet minimization computes the strongest bound given all load dimensions. To test it, a `Client.replace` method has been added, which has some limitations (see implementation).~

Some open questions remain:
- Interface:
  - ~do we want backwards compatibility (like in this PR) or do we want to enforce passing and returning a list/vector always?~
  - ~should the default be 1 or 0 (no) load dimensions? Currently we support 0 load dimensions but default to 1. When using 0 load dimensions, the relevant properties should not be accessed.~
- Performance:
  - is the overhead acceptable given the new functionality?
  - should we introduce seperate penalties per load dimension and manage them per dimension? If, so we should probably interpret `target_feasible` as the desired percentage feasible for all load dimensions, not per dimension independently.
  - in this PR, computing `SegmentBetween` for a single load dimension takes `O(n)` where `n` is the length of the segment. This may become significant with many load dimensions (e.g. 64) which can often be avoided by first checking in `O(1)` if the segment contains any pickups, and if not, simply compute the difference in cumulative loads (see details). Compiling out P&D altogether when not needed is even better in our experience.
<details>

```cpp
LoadSegment Route::SegmentBetween::load(size_t dimension) const
{
    assert(start > 0);  // SegmentBetween should not start at depot

    LoadSegment const &startLoadSegment
        = route->loadBefore[start - 1][dimension];
    LoadSegment const &endLoadSegment = route->loadBefore[end][dimension];

    // No pickups, so LoadSegment is simply the difference in deliveries
    // between end and start of the segment.
    if (startLoadSegment.pickup() == endLoadSegment.pickup())
    {
        Load const load
            = endLoadSegment.delivery() - startLoadSegment.delivery();
        assert(load >= 0);
        return LoadSegment(load, 0, load);
    }

    auto loadSegment = route->loadAt[start][dimension];
    for (size_t step = start; step != end; ++step)
        loadSegment = LoadSegment::merge(loadSegment,
                                         route->loadAt[step + 1][dimension]);

    return loadSegment;
}
```
</details>

## Benchmarks on VRPTW

Mostly to check for regressions: [here](https://github.com/PyVRP/PyVRP/pull/645#issuecomment-2442225406).

### CODE REVIEW

**Comment by N-Wouda** [comment]
Our terminology for this (`pickup`, `delivery`, and `capacity` on the VehicleType) already aligns with VROOM, so I think we should not update that.

**Comment by N-Wouda** [comment]
Some learnings from this so far:
- We need to bind a custom init method to also allow the common case of a single load dimension. Internally we can work with vectors. This'll keep everything backwards compatible, and has the added benefit of not requiring any updates to the existing tests.
- Validation and padding etc. work well. Let's keep that.
- I am not sure about how to extend `LoadSegment`, and @wouterkool indicated the use of `std::vector` in the current implementation might be problematic. Makes sense. I'll wait for his alternative implementation to see how we can combine things and keep it fast.

It's probably easiest to start from a new branch now.

**Comment by N-Wouda** [comment]
I've updated the interface a bit and simplified where possible to get us to just one default (all lists). I haven't looked at the LS evaluation yet, that's for later.

**Comment by N-Wouda** [comment]
I should have sufficient time next week to (finally!) finish this. 🤞

**Comment by N-Wouda** [comment]
> * should we introduce seperate penalties per load dimension and manage them per dimension? If, so we should probably interpret `target_feasible` as the desired percentage feasible for all load dimensions, not per dimension independently.

I think we can do without for now, but we'll need to run some experiments to determine if this is OK. I know of no real multidim instances from the literature, but we can use multiple load dims also to model site-dependent VRP. Those instances can probably serve as an initial benchmark.

> * in this PR, computing `SegmentBetween` for a single load dimension takes `O(n)` where `n` is the length of the segment. This may become significant with many load dimensions (e.g. 64) which can often be avoided by first checking in `O(1)` if the segment contains any pickups, and if not, simply compute the difference in cumulative loads (see details). Compiling out P&D altogether when not needed is even better in our experience.

I think we can leave this to a follow-up PR, assuming performance is not too bad right now.

**Comment by N-Wouda** [comment]
I'm going to have one more look in a minute. I think this is otherwise good to go and ready for some (performance) tests to ensure there's no significant regressions w.r.t. current main.

**Comment by N-Wouda** [comment]
LGTM now. This needs a performance check, so I need to figure out how the HPC stuff works in Rotterdam first. I'll have a look at that tomorrow.

**Comment by N-Wouda** [comment]
@leonlan @wouterkool could you both/either of you have a look at this?

**Comment by leonlan** [comment]
I'll probably have time on Wednesday to review this.

**Comment by leonlan** [comment]
GH1000, 60 min, 10 seeds
- main https://github.com/PyVRP/PyVRP/commit/452289225fe28ec6b04f7aac9aee3e253b31b462 - gap: 0.74%, iters: 53172
- latest commit https://github.com/PyVRP/PyVRP/pull/645/commits/ffde097214f10ab75e0bcffe3bbe054041e72ba3 - gap: 0.75%, iters: 51091

**Comment by N-Wouda** [comment]
> GH1000, 60 min, 10 seeds
> 
> * main [4522892](https://github.com/PyVRP/PyVRP/commit/452289225fe28ec6b04f7aac9aee3e253b31b462) - gap: 0.74%, iters: 53172
> * latest commit [ffde097](https://github.com/PyVRP/PyVRP/commit/ffde097214f10ab75e0bcffe3bbe054041e72ba3) - gap: 0.75%, iters: 51091

Thanks! That's some 4% worse in terms of iterations, but similar quality. That seems to be roughly in line with the fact there's simply more work that need to be done in this PR. Should be fine!

**Comment by N-Wouda** [comment]
Seems rather arbitrary 😆

**Review by leonlan — COMMENTED** [review]
Looks good to me! I think @wouterkool should also review it since he has worked on it.

> - is the overhead acceptable given the new functionality?

I think 4% less iterations is fine since it does not really impact the solution gap. 

> - should we introduce seperate penalties per load dimension and manage them per dimension? If, so we should probably interpret target_feasible as the desired percentage feasible for all load dimensions, not per dimension independently.

I think we should postpone this until we have some instances/examples where it matters.
    
> - in this PR, computing SegmentBetween for a single load dimension takes O(n) where n is the length of the segment. This may become significant with many load dimensions (e.g. 64) which can often be avoided by first checking in O(1) if the segment contains any pickups, and if not, simply compute the difference in cumulative loads (see details). Compiling out P&D altogether when not needed is even better in our experience.

This also relates to #555 right? I don't know how much it's actually faster, this kind of relates to the lack of examples/use cases like the previous point.

File: pyvrp/cpp/ProblemData.cpp [hunk 1]
```diff
@@ -528,7 +562,15 @@ ProblemData::ProblemData(std::vector<Client> clients,
...
-                                   { return sum + type.numAvailable; }))
+                                   { return sum + type.numAvailable; })),
+      numLoadDimensions_(
+          clients_.empty()
+              // If there are no clients we look at the vehicle types. If both
+              // are empty we default to 0. Clients have pickups and deliveries
+              // but the client constructor already ensures those are of equal
+              // size (within a single client).
+              ? (vehicleTypes_.empty() ? 0 : vehicleTypes_[0].capacity.size())
+              : clients_[0].delivery.size())
```
[the remark this answers is not in the document]
Comment (N-Wouda) [reply pyvrp/cpp/ProblemData.cpp:573]: Validation happens after this, so accessing `vehicleTypes_[0]` here may be out of bounds without the check.

**Comment by leonlan** [comment]
@wouterkool do you have an estimate of when you could review this PR?

File: pyvrp/plotting/plot_route_schedule.py [hunk 2]
```diff
@@ -148,16 +157,20 @@ def add_traces(dist, t, drive_time, serv_time, load):
     )
 
     # Plot remaining load on second axis
+    capacity = 0
+    if track_load:
+        capacity = vehicle_type.capacity[load_dimension]
+
     twin1 = ax.twinx()
     twin1.fill_between(
```
[the remark this answers is not in the document]
Comment (N-Wouda) [reply pyvrp/plotting/plot_route_schedule.py:165]: Works for me

**Pull request merged**